src/core/bbr.c

```diff
@@ -0,0 +1,1327 @@
+
+    if (++Bbr->SlowStartupRoundCounter >= kStartupSlowGrowRoundLimit) {
+        Bbr->BtlbwFound = TRUE;
+    }
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+void
+BbrCongestionControlTransitToProbeRtt(
+    _In_ QUIC_CONGESTION_CONTROL* Cc,
+    _In_ uint64_t LargestPacketNumberSent
+    )
+{
+    QUIC_CONGESTION_CONTROL_BBR* Bbr = &Cc->Bbr;
+
+    Bbr->BbrState = BBR_STATE_PROBE_RTT;
+    Bbr->PacingGain = BBR_UNIT;
+    Bbr->EarliestTimeToExitProbeRttValid = FALSE;
+    Bbr->ProbeRttRoundValid = FALSE;
+
+    BbrBandwidthSamplerOnAppLimited(&Bbr->BandwidthSampler, LargestPacketNumberSent);
+
+    Bbr->AppLimitedSinceProbeRtt = FALSE;
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+void
+BbrCongestionControlTransitToDrain(
+    _In_ QUIC_CONGESTION_CONTROL* Cc
+    )
+{
+    QUIC_CONGESTION_CONTROL_BBR* Bbr = &Cc->Bbr;
+
+    Bbr->BbrState = BBR_STATE_DRAIN;
+    Bbr->PacingGain = kDrainGain;
+    Bbr->CwndGain = kStartupGain;
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+BOOLEAN
+BbrCongestionControlShouldExitDrain(
+    _In_ QUIC_CONGESTION_CONTROL* Cc
+    )
+{
+    return Cc->Bbr.BbrState == BBR_STATE_DRAIN &&
+           Cc->Bbr.BytesInFlight <= BbrCongestionControlCalculateTargetCwnd(Cc, BBR_UNIT);
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+BOOLEAN
+BbrCongestionControlShouldProbeRtt(
+    _In_ QUIC_CONGESTION_CONTROL* Cc
+    )
+{
+    QUIC_CONGESTION_CONTROL_BBR *Bbr = &Cc->Bbr;
+    if (Bbr->BbrState != BBR_STATE_PROBE_RTT
+        && !Bbr->ExitingQuiescene
+        && Bbr->MinRttSampler.RttSampleExpired) {
+
+        return TRUE;
+
+    }
+    return FALSE;
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+void
+BbrCongestionControlUpdateCongestionWindow(
+    _In_ QUIC_CONGESTION_CONTROL* Cc,
+    _In_ uint64_t TotalBytesAcked,
+    _In_ uint64_t AckedBytes
+    )
+{
+    QUIC_CONGESTION_CONTROL_BBR *Bbr = &Cc->Bbr;
+    QUIC_CONNECTION* Connection = QuicCongestionControlGetConnection(Cc);
+
+    if (Bbr->BbrState == BBR_STATE_PROBE_RTT) {
+        return;
+    }
+
+    uint64_t Bandwidth = BbrCongestionControlGetBandwidth(Cc);
+
+    uint64_t PacingRate = Bandwidth * Bbr->PacingGain / BBR_UNIT;
+
+    if (PacingRate < kLowPacingRateForSendQuantumBytesPerSecond * BW_UNIT) {
+        Bbr->SendQuantum = Connection->Paths[0].Mtu;
+    } else if (PacingRate < kHighPacingRateForSendQuantumBytesPerSecond * BW_UNIT) {
+        Bbr->SendQuantum = Connection->Paths[0].Mtu * 2;
+    } else {
+        Bbr->SendQuantum = CXPLAT_MIN(PacingRate * kMilliSecsInSec / BW_UNIT, k64K);
+    }
+
+    uint64_t TargetCwnd = BbrCongestionControlCalculateTargetCwnd(Cc, Bbr->CwndGain);
+    if (Bbr->BtlbwFound) {
+        TargetCwnd += WindowedFilterGetBest(&Bbr->MaxAckHeightFilter);
+    }
+    
+    uint32_t CongestionWindow = Bbr->CongestionWindow;
+
+    if (Bbr->BtlbwFound) {
+        CongestionWindow = (uint32_t)CXPLAT_MIN(TargetCwnd, CongestionWindow + AckedBytes);
+    } else if (CongestionWindow < TargetCwnd || TotalBytesAcked < Bbr->InitialCongestionWindow) {
```

Comment:
It seems like this line here is the only reason the TotalBytesAcked field was added to QUIC_LOSS_DETECTION. Can't we just have BbrCongestionControlCalculateTargetCwnd always return a window >= InitialCongestionWindow? That would make this extra "TotalBytesAcked < Bbr->InitialCongestionWindow" logic unnecessary.

Reply:
I think it's different, see [Core cwnd Adjustment Mechanism](https://tools.ietf.org/id/draft-cardwell-iccrg-bbr-congestion-control-00.html#core-cwnd-adjustment-mechanism) part in the BBR doc.

I guess your focus is to remove the extra `TotalBytesAcked` in `QUIC_LOSS_DETECTION`. If it's true, we can move that field  (other fields as well) to BBR to reduce the overhead for other CC algos.